### PR TITLE
Add v1.31, v1.30 to GHA matrix and use latest Kind k8s patch for each minor versions for e2e

### PR DESCRIPTION
**Pull request by kaovilai**
Always test latest available patch version of each supported k8s version available in Kindest/node images.

ie. This adds v1.31, v1.30 to test matrix and upgrade patch versions for others. Removes v1.23 and v1.24.

```diff
        k8s:
-          - 1.23.17
-          - 1.24.17
           - 1.25.16
-          - 1.26.13
+          - 1.26.15
-          - 1.27.10
+          - 1.27.16
-          - 1.28.6
+          - 1.28.13
-          - 1.29.1
+          - 1.29.8
+          - 1.30.4
+          - 1.31.1
```

Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Fixes #(issue)

# Please indicate you've done the following:

- [ ] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file (`make new-changelog`)](https://velero.io/docs/main/code-standards/#adding-a-changelog) or comment `/kind changelog-not-required` on this PR.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

**Comment by kaovilai** [comment]
issues:
1.25.9 taken precedence over 1.25.1[0-9] ie. 1.25.17

**Comment by kaovilai** [comment]
moving `sort` to after `awk` fixes it. awk first remove version duplicates where the later ones are kept. then sort can order the remaining minor versions.

File: .github/workflows/crds-verify-kind.yaml [hunk 1]
```diff
@@ -35,21 +35,29 @@ jobs:
...
+      GH_TOKEN: ${{ github.token }}
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+    steps:
+      - name: Set k8s versions
+        id: set-matrix
+        # everything excluding older tags below 1.23. limits needs to be high enough to cover all latest versions
+        # grep -E "v[1-9]\.(2[3-9]|[3-9][0-9])" filters for v1.23 to v9.99
+        run: |
+          echo "matrix={\"k8s\":$(wget -q -O - "https://hub.docker.com/v2/namespaces/kindest/repositories/node/tags?page_size=50" | grep -o '"name": *"[^"]*' | grep -o '[^"]*$' | grep -v -E "alpha|beta" | grep -E "v[1-9]\.(2[3-9]|[3-9][0-9])" | awk -F. '{if(!a[$1"."$2]++)print $1"."$2"."$NF}' | sort -r | sed s/v//g | jq -R -c -s 'split("\n")[:-1]')}" >> $GITHUB_OUTPUT
```
Comment (blackpiglet) [inline .github/workflows/crds-verify-kind.yaml:52]: Could you give more explanation and comments to this piece of code ` awk -F. '{if(!a[$1"."$2]++)print $1"."$2"."$NF}'`?

It's not easy to understand what it does.

Comment (blackpiglet) [reply .github/workflows/crds-verify-kind.yaml:52]: Another comment is we don't need to support that much of the KinD versions.
I suggest to deprecate the testing of v1.23 and v1.24.

Comment (kaovilai) [reply .github/workflows/crds-verify-kind.yaml:52]: The awk removes older patches of the same release so only 1.30.1 runs, and not 1.30.0

File: .github/workflows/e2e-test-kind.yaml [hunk 2]
```diff
@@ -43,26 +43,40 @@ jobs:
...
+  setup-test-matrix:
+    runs-on: ubuntu-latest
+    env:
+      GH_TOKEN: ${{ github.token }}
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+    steps:
+      - name: Set k8s versions
+        id: set-matrix
+        # everything excluding older tags below 1.23. limits needs to be high enough to cover all latest versions
```
Comment (mateusoliveira43) [inline .github/workflows/e2e-test-kind.yaml:57]: here (and PR description) needs update that lowest version is 1.25.x?

File: .github/workflows/crds-verify-kind.yaml
Comment (mateusoliveira43) [inline .github/workflows/crds-verify-kind.yaml:1]: can be a separate PR, but this file and `.github/workflows/e2e-test-kind.yaml` have a lot of duplication 

It would be better to merge then together?

**Review by mateusoliveira43 — APPROVED** [review]

File: .github/workflows/crds-verify-kind.yaml
Comment (kaovilai) [reply .github/workflows/crds-verify-kind.yaml:1]: that's an opinionated change over existing content of this pr. I think that can be discussed/opened separately.

Comment (kaovilai) [reply .github/workflows/crds-verify-kind.yaml:1]: IIUC not combining jobs, just the workflow file right?

Comment (mateusoliveira43) [reply .github/workflows/crds-verify-kind.yaml:1]: Simplest change I think would be copying `jobs.crd-check` step to `.github/workflows/e2e-test-kind.yaml`

that would run pre steps (`build-cli` and `setup-test-matrix`) only once and `crd-check` and `run-e2e-test` would still run in parallel

Comment (kaovilai) [reply .github/workflows/crds-verify-kind.yaml:1]: +1

**Review by blackpiglet — APPROVED** [review]

**Review by shubham-pampattiwar — APPROVED** [review]
